### PR TITLE
Invalidateapi

### PR DESCRIPTION
**Pull request by azidar**
DO NOT MERGE.

### CODE REVIEW

**Comment by ucbjrl** [comment]
Here's the proposed [API document](https://docs.google.com/a/berkeley.edu/document/d/1EzhXL5EMw8vQMNppmqVNJC4qVyEST7tKzW9YBNamIOo/edit?usp=sharing)

File: src/test/scala/chiselTests/package.scala [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import chisel3.core.LegacyModule
+
+package object chiselTests {
+  // We don't want firrtl complaining about "not fully initialized" connections.
+  val implicitInvalidateOptions = chisel3.core.ExplicitCompileOptions.Strict.copy(explicitInvalidate = false)
+  abstract class ImplicitInvalidateModule extends LegacyModule()(implicitInvalidateOptions)
```
Comment (ducky64) [inline src/test/scala/chiselTests/package.scala]: If this is breaking enough tests that you don't want to fix all the cases correctly and need to create an abstraction to patch around existing issues, you might want to rethink the API

File: src/test/scala/chiselTests/InvalidateAPISpec.scala [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+import org.scalatest._
+
+class InvalidateAPISpec extends ChiselPropSpec with Matchers {
+
+  def myGenerateFirrtl(t: => Module): String = Driver.emit(() => t)
+  def compileFirrtl(t: => Module): Unit = {
+    Driver.execute(Array[String]("--compiler", "verilog"), () => t)
+  }
+
+  class HardIP extends Bundle {
```
Comment (ducky64) [inline src/test/scala/chiselTests/InvalidateAPISpec.scala]: Name isn't descriptive

File: src/test/scala/chiselTests/InvalidateAPISpec.scala [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    val in  = Input(Bool())
+    val out = Output(Bool())
+
+    override def cloneType: this.type = (new HardIP).asInstanceOf[this.type]
```
Comment (ducky64) [inline src/test/scala/chiselTests/InvalidateAPISpec.scala]: Is this necessary? This bundle takes no parameters

File: src/test/scala/chiselTests/InvalidateAPISpec.scala [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  }
+
+  property("an output connected to DontCare should emit a Firrtl \"is invalid\" with Strict CompileOptions") {
+    import chisel3.core.ExplicitCompileOptions.Strict
+    class ModuleWithDontCare extends Module {
+
```
Comment (ducky64) [inline src/test/scala/chiselTests/InvalidateAPISpec.scala]: Style: why is there a blank line here?
(and ditto for all below)

File: src/test/scala/chiselTests/InvalidateAPISpec.scala [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+  property("a DontCare cannot be a connection sink (LHS) for := ") {
+    import chisel3.core.ExplicitCompileOptions.Strict
+    class ModuleWithDontCareSink extends Module {
+
+      val io = IO(new HardIP)
+      DontCare := io.in
+    }
+
+    val exception = intercept[ChiselException] {
```
Comment (ducky64) [inline src/test/scala/chiselTests/InvalidateAPISpec.scala]: alternatively:
```scala
a [ChiselException] should be thrownBy {
  ...
}
```

File: chiselFrontend/src/main/scala/chisel3/core/UserModule.scala [hunk 6]
```diff
@@ -97,8 +103,11 @@ abstract class ImplicitModule(implicit moduleCompileOptions: CompileOptions)
   private[core] override def initializeInParent() {
     implicit val sourceInfo = UnlocatableSourceInfo
 
-    for (port <- getModulePorts) {
-      pushCommand(DefInvalid(sourceInfo, port.ref))
+    // Generate firrtl invalids if the client doesn't want explicit control over their generation.
```
Comment (ucbjrl) [inline chiselFrontend/src/main/scala/chisel3/core/UserModule.scala]: I'm not sure. I'll see if I can verify this.

**Comment by ucbjrl** [comment]
I've tried to convert to a binding-based implementation. We still have the `DontCare` object and a new `DontCareElement`, but the implementation has been pushed down into the general connection logic.

There are more changes then the previous attempt. I'm not happy with the direct tests against the `DontCare` singleton - I tried using a match on `binding` but that has to be guarded with a `.hasBinding` test which results in duplication of the normal (non-`DontCare`) code.

I have eliminated the hacks to individual tests to deal with the FIRRTL `.when/.otherwise` issue - no longer required since we've set the current `CompileOptions.Strict.explicitInvalidate` to false, so we get the current implicit invalidate behavior in all _normal_ cases.

File: src/test/scala/chiselTests/VectorPacketIO.scala [hunk 7]
```diff
@@ -40,6 +41,8 @@ class BrokenVectorPacketModule extends Module {
   val n  = 4
   val io = IO(new VectorPacketIO(n))
 
+  io := DontCare
```
Comment (ducky64) [inline src/test/scala/chiselTests/VectorPacketIO.scala]: Why is this necessary?

File: chiselFrontend/src/main/scala/chisel3/core/UserModule.scala [hunk 8]
```diff
@@ -67,9 +67,15 @@ abstract class UserModule(implicit moduleCompileOptions: CompileOptions)
...
-    // Generate IO invalidation commands to initialize outputs as unused
-    val invalidateCommands = getModulePorts map {port => DefInvalid(UnlocatableSourceInfo, port.ref)}
-
+    // Generate IO invalidation commands to initialize outputs as unused,
+    //  unless the client wants explicit control over their generation.
+    val invalidateCommands = {
+      if (!compileOptions.explicitInvalidate) {
+        getModulePorts map { port => DefInvalid(UnlocatableSourceInfo, port.ref) }
+      } else {
+        Seq()
```
[the remark this answers is not in the document]
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/UserModule.scala]: I don't understand your comment here. This is the fundamental feature of the InvalidateAPI: if enabled, user code must explicitly generate "is invalid" using DontCare connections, otherwise, we generate blanket "is invalid" - the current (pre-InvalidateAPI) behavior. This is where we deal with Module IO ports.

File: chiselFrontend/src/main/scala/chisel3/core/UserModule.scala [hunk 9]
```diff
@@ -97,10 +103,6 @@ abstract class ImplicitModule(implicit moduleCompileOptions: CompileOptions)
   private[core] override def initializeInParent() {
     implicit val sourceInfo = UnlocatableSourceInfo
 
-    for (port <- getModulePorts) {
```
[the remark this answers is not in the document]
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/UserModule.scala]: This just eliminates redundant/duplicated code. explicitInvalidate is dealt with in the parent class.

File: src/test/scala/chiselTests/ComplexAssign.scala [hunk 10]
```diff
@@ -47,6 +48,9 @@ class ComplexAssignTester(enList: List[Boolean], re: Int, im: Int) extends Basic
 
 class ComplexAssignSpec extends ChiselPropSpec {
   property("All complex assignments should return the correct result") {
+    // Disable shrinking on error.
```
[the remark this answers is not in the document]
Comment (ucbjrl) [reply src/test/scala/chiselTests/ComplexAssign.scala]: This is an aid to debugging failing tests. It isn't strictly required for this PR, but was helpful in debugging failures. It can be removed.

File: chiselFrontend/src/main/scala/chisel3/core/UserModule.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (ducky64) [reply chiselFrontend/src/main/scala/chisel3/core/UserModule.scala]: Ok, so this is kind of a compatibility switch for previous code that didn't have this in mind.

File: src/test/scala/chiselTests/FixedPointSpec.scala [hunk 11]
```diff
@@ -108,6 +109,7 @@ class SBPTester extends BasicTester {
   assert(dut.io.out === 3.0.F(0.BP))
 
   val test = Wire(FixedPoint(10.W, 5.BP))
+  test := DontCare
```
[the remark this answers is not in the document]
Comment (ucbjrl) [reply src/test/scala/chiselTests/FixedPointSpec.scala]: This suppresses the "Reference test is not fully initialized" error from firrtl. It seemed the minimum change to keep the test working.

File: src/test/scala/chiselTests/VectorPacketIO.scala (hunk 7 above)
Comment (ucbjrl) [reply src/test/scala/chiselTests/VectorPacketIO.scala]: This suppresses the "Reference io is not fully initialized" error from firrtl. It seemed the minimum change to keep the test working.

File: build.sbt [hunk 12]
```diff
@@ -71,10 +71,6 @@ lazy val chiselSettings = Seq (
         <distribution>repo</distribution>
       </license>
     </licenses>
-    <scm>
```
[the remark this answers is not in the document]
Comment (ucbjrl) [reply build.sbt]: It isn't required for this PR. It's been rolled back.

File: chiselFrontend/src/main/scala/chisel3/core/BiConnect.scala [hunk 13]
```diff
@@ -89,6 +89,33 @@ object BiConnect {
         case _ => recordConnect(sourceInfo, connectCompileOptions, left_r, right_r, context_mod)
       }
 
+        // Right (source) is a DontCare
```
[the remark this answers is not in the document]
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/BiConnect.scala]: Corrected indentation.

File: chiselFrontend/src/main/scala/chisel3/core/BiConnect.scala [hunk 14]
```diff
@@ -89,6 +89,33 @@ object BiConnect {
...
+      case (left_e: Element, DontCare) =>
+        elemConnect(sourceInfo, connectCompileOptions, left_e, DontCareElement, context_mod)
+      // // Handle Vec case - connect to DontCare. Apply the DontCare to each element of the Vec.
```
[the remark this answers is not in the document]
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/BiConnect.scala]: Corrected indentation.

File: chiselFrontend/src/main/scala/chisel3/core/BiConnect.scala [hunk 15]
```diff
@@ -89,6 +89,33 @@ object BiConnect {
...
+        for((field, left_sub) <- left_r.elements) {
+          try {
+            connect(sourceInfo, connectCompileOptions, left_sub, DontCare, context_mod)
+          } catch {
+            case BiConnectException(message) => throw BiConnectException(s".$field$message")
+          }
+        }
+
+      // DontCare cannot be a sink (LHS)
+      case (DontCare, _) => throw DontCareCantBeSink
```
[the remark this answers is not in the document]
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/BiConnect.scala]: Excellent suggestion. I've moved the bulk of the logic to elemConnect(). I think the implementation is much cleaner. Thanks @ducky64

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala [hunk 16]
```diff
@@ -430,3 +432,80 @@ object WireInit {
     x
   }
 }
+
+/** RHS (source) for Invalidate API.
+  * Causes connection logic to emit a DefInvalid when connected to an output port (or wire).
+  */
+object DontCare extends Data {
+  binding = DontCareBinding()
+  direction = ActualDirection.Output
```
[the remark this answers is not in the document]
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Simplified. I believe the answer is largely in the affirmative.

File: chiselFrontend/src/main/scala/chisel3/core/UserModule.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (ducky64) [reply chiselFrontend/src/main/scala/chisel3/core/UserModule.scala]: Ok, and having the DefInvalidate in the parent class handles both cases where the parent doesn't drive inputs and where the instance doesn't drive outputs?

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala [hunk 17]
```diff
@@ -430,3 +432,80 @@ object WireInit {
...
+  def cloneType = DontCare
+  // Define setter/getter pairing
+  // DontCare can only be connected to Ports and Wires (and Unbound), and only as a source
+  // Attempts to change its binding are silently ignore.
```
[the remark this answers is not in the document]
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Attempt to change the binding should now generate errors.

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala [hunk 18]
```diff
@@ -430,3 +432,80 @@ object WireInit {
...
+  private[chisel3] override def bind(target: Binding, parentDirection: SpecifiedDirection) {
```
[the remark this answers is not in the document]
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Agreed.

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala [hunk 19]
```diff
@@ -430,3 +432,80 @@ object WireInit {
...
+    val resolvedDirection = SpecifiedDirection.fromParent(parentDirection, specifiedDirection)
+    direction = resolvedDirection match {
+      case SpecifiedDirection.Unspecified | SpecifiedDirection.Output => ActualDirection.Output
+      case SpecifiedDirection.Input => throwException(s"DontCare may not be an input")
+    }
+  }
+
+  override private[chisel3] def flatten : IndexedSeq[Element] = IndexedSeq.empty[Element]
```
[the remark this answers is not in the document]
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Eliminated.

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala [hunk 20]
```diff
@@ -430,3 +432,80 @@ object WireInit {
...
+  private[chisel3] def toType = "DontCare"
+  private[chisel3] def allElements: Seq[Element] = Seq.empty[Element]
+  private[core] def legacyConnect(that: Data)(implicit sourceInfo: SourceInfo): Unit = {
+    Builder.error("legacyConnect: DontCare cannot be a connection sink (LHS)")
+  }
+  private[core] def cloneTypeWidth(width: Width): this.type = cloneType
```
[the remark this answers is not in the document]
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Make DontCare extend Element.

File: src/test/scala/chiselTests/ComplexAssign.scala (hunk 10 above)
[the remark this answers is not in the document]
Comment (ducky64) [reply src/test/scala/chiselTests/ComplexAssign.scala]: Sure, keep it in if you think it's overall generalizable (the philosophy here is generally don't leave your debugging printfs sprinkled throughout the codebase)

File: src/main/scala/chisel3/internal/firrtl/Emitter.scala [hunk 21]
```diff
@@ -160,9 +160,13 @@ private class Emitter(circuit: Circuit) {
     * enclosing block upon emission.
     */
   private def processWhens(cmds: Seq[Command]):
-  Seq[Command] = { cmds.zip(cmds.tail).map({ case (a: WhenEnd, b:
-  AltBegin) => a.copy(hasAlt = true) case (a, b) => a }) ++
-  cmds.lastOption }
+  Seq[Command] = { if (cmds.isEmpty) {
```
[the remark this answers is not in the document]
Comment (ucbjrl) [reply src/main/scala/chisel3/internal/firrtl/Emitter.scala]: It used to be the case that cmds could never be empty: modules must have IO, and we generated "is invalid" for those IOs whether they were connected or not. With the InvalidateAPI, this is no longer the case and cmds may be empty. We need to check for this.

File: src/test/scala/chiselTests/FixedPointSpec.scala (hunk 11 above)
[the remark this answers is not in the document]
Comment (ducky64) [reply src/test/scala/chiselTests/FixedPointSpec.scala]: Does it make more sense to assign a literal to it instead of also (unintentionally) testing DontCare behavior?

File: src/test/scala/chiselTests/VectorPacketIO.scala (hunk 7 above)
Comment (ducky64) [reply src/test/scala/chiselTests/VectorPacketIO.scala]: Can we assign something to it so this doesn't also test DontCare behavior?

File: chiselFrontend/src/main/scala/chisel3/core/UserModule.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/UserModule.scala]: Correct.

File: chiselFrontend/src/main/scala/chisel3/core/UserModule.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/UserModule.scala]: Hmmm. I'm not sure I understand this question. I meant the logic in ImplicitModule is a duplicate of logic in the parent class (UserModule.generateComponent()). If ImplicitModule overrode generateComponent() and didn't invoke its parent's implementation, then we'd need the duplicate. Am I missing something more subtle here?

File: src/main/scala/chisel3/internal/firrtl/Emitter.scala (hunk 21 above)
[the remark this answers is not in the document]
Comment (ducky64) [reply src/main/scala/chisel3/internal/firrtl/Emitter.scala]: Ok, I see.
Second question: why is this written in such a unreadable manner? (sure, it was originally that way, but unless there's a good reason it needs to look like a terrible text editor tried to 80-col wrap someone's cutesy one-liner, this should be fixed...)

File: chiselFrontend/src/main/scala/chisel3/core/BiConnect.scala [hunk 22]
```diff
@@ -89,6 +89,12 @@ object BiConnect {
         case _ => recordConnect(sourceInfo, connectCompileOptions, left_r, right_r, context_mod)
       }
 
+      // Right is DontCare - it may be connected to anything. It generates a defInvalid for the left.
+      case (left, DontCare) => pushCommand(DefInvalid(sourceInfo, left.lref))
```
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/BiConnect.scala]: Again, is BiConnect meant to be directional? Why does it matter is DontCare is on the left or right side?

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala [hunk 23]
```diff
@@ -430,3 +435,34 @@ object WireInit {
     x
   }
 }
+
+/** RHS (source) for Invalidate API.
+  * Causes connection logic to emit a DefInvalid when connected to an output port (or wire).
+  */
+object DontCare extends Element(width = UnknownWidth()) {
+  binding = DontCareBinding()
```
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Consider just calling Data/Element's binding function instead of messing with internal variables. That also prevents rebinding, unless there's a compelling reason to support it

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala [hunk 24]
```diff
@@ -430,3 +435,34 @@ object WireInit {
...
+  private[core] def connectFromBits(that: chisel3.core.Bits)(implicit sourceInfo:  SourceInfo, compileOptions: CompileOptions): Unit = {
+    Builder.error("connectFromBits: DontCare cannot be a connection sink (LHS)")
+  }
+
+  def do_asUInt(implicit sourceInfo: chisel3.internal.sourceinfo.SourceInfo, compileOptions: CompileOptions): chisel3.core.UInt = {
+    Builder.error("DontCare does not have a UInt representation")
+    0.U
+  }
+  // DontCare's match any type.
+  private[core] def typeEquivalent(that: chisel3.core.Data): Boolean = true
```
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Why?
I think it makes more sense for DontCares to not be equivalent to anything but another DoneCare, or to have this crash.

File: chiselFrontend/src/main/scala/chisel3/core/MonoConnect.scala [hunk 25]
```diff
@@ -136,6 +145,9 @@ object MonoConnect {
         //    CURRENT MOD   CURRENT MOD
         case (Output,       _) => issueConnect(sink, source)
         case (Internal,     _) => issueConnect(sink, source)
+        // Normally, this (an Input on the LHS of a connection) should generate an error,
+        //  but it's the way we indicate "DontCare"s
+        case (Input, Internal) if source.binding == DontCareBinding() => issueConnect(sink, source)
```
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/MonoConnect.scala]: Wait, does this make sense? You can assign a DontCare to an input? Why should a module be writing its inputs?

File: chiselFrontend/src/main/scala/chisel3/core/MonoConnect.scala [hunk 26]
```diff
@@ -107,14 +109,21 @@ object MonoConnect {
             case MonoConnectException(message) => throw MonoConnectException(s".$field$message")
           }
         }
-
+      // Source is DontCare - it may be connected to anything. It generates a defInvalid for the sink.
+      case (sink, DontCare) => pushCommand(DefInvalid(sourceInfo, sink.lref))
```
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/MonoConnect.scala]: Is this being duplicated with the logic in issueConnect? Abstraction level wise this seems out of place with the rest of the cases here...

File: chiselFrontend/src/main/scala/chisel3/core/UserModule.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (ducky64) [reply chiselFrontend/src/main/scala/chisel3/core/UserModule.scala]: So what happens if a module doesn't assign anything to its output? The DefInvalid shouldn't be set in the parent class since it would be trying to write the output of a child module.

File: chiselFrontend/src/main/scala/chisel3/core/BiConnect.scala (hunk 22 above)
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/BiConnect.scala]: Right you are.

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala (hunk 23 above)
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Done.

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala (hunk 24 above)
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Done.

File: chiselFrontend/src/main/scala/chisel3/core/MonoConnect.scala (hunk 26 above)
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/MonoConnect.scala]: No, it deals with the case where we connect DontCares to Vecs or Bundles. It eliminates the logic that had to pull those aggregates apart to get down to their leaves. This is closer to the way the current code (non-invalidateapi) deals with "is invalid".

File: chiselFrontend/src/main/scala/chisel3/core/MonoConnect.scala (hunk 25 above)
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/MonoConnect.scala]: I agree it looks screwy, but chisel3 currently generates "is invalid" for inputs and this is an attempt to allow generation of that explicitly.

**Comment by ucbjrl** [comment]
Sorry, I think there's a nomenclature issue here (eliminating redundant defInvalid code in ImplicitModule). It has nothing to do with chisel modules. There was duplication of code in ImplicitModule and its super class, UserModule.

File: chiselFrontend/src/main/scala/chisel3/core/MonoConnect.scala (hunk 26 above)
Comment (ducky64) [reply chiselFrontend/src/main/scala/chisel3/core/MonoConnect.scala]: Do we want to pull apart vecs and bundles as with the rest of the code? for example, what happens if you invalidate a bidirectional bundle?

File: chiselFrontend/src/main/scala/chisel3/core/MonoConnect.scala (hunk 25 above)
Comment (ducky64) [reply chiselFrontend/src/main/scala/chisel3/core/MonoConnect.scala]: is that something we want to keep supporting? can we get rid of that?

Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/MonoConnect.scala]: I don't feel qualified enough to comment, but we should settle this before merging.

File: chiselFrontend/src/main/scala/chisel3/core/MonoConnect.scala (hunk 26 above)
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/MonoConnect.scala]: Good question. If we do pull apart aggregates (for the purpose of DontCare connections), firrtl could be more helpful reporting which elements of an aggregate aren't completely initialized. I think this is an implementation issue which could be resolved post release - it shouldn't change the API.

**Comment by ucbjrl** [comment]
I've updated the [API document](https://docs.google.com/a/berkeley.edu/document/d/1EzhXL5EMw8vQMNppmqVNJC4qVyEST7tKzW9YBNamIOo/edit?usp=sharing) with an issue that has emerged while testing existing code with the new API.

Specifically, this sequence:
```scala
  ...
  when(reset.toBool()) {
    io.out := false.B
  }

  when(io.in) {
    io.out := true.B
  }
  ...
```
will provoke a _Reference io is not fully initialized._ complaint from firrtl.

File: chiselFrontend/src/main/scala/chisel3/core/MonoConnect.scala (hunk 25 above)
Comment (ducky64) [reply chiselFrontend/src/main/scala/chisel3/core/MonoConnect.scala]: This sounds related to firrtl warnings for partially initialized bundles, especially if they'r ebidirectional

**Comment by jackkoenig** [comment]
For the record, I don't think `DontCare` should really be in package experimental if your old code now breaks and you need to use this new thing.

**Comment by ucbjrl** [comment]
@jackkoenig raises a good point. We should either move `DontCare` out of experimental, or update CompileOptions so the default behavior does not require `explicitInvalidate`.

File: chiselFrontend/src/main/scala/chisel3/core/BiConnect.scala [hunk 27]
```diff
@@ -89,9 +102,40 @@ object BiConnect {
...
+            // For each field in left, descend with right
+            for ((field, left_sub) <- left_r.elements) {
+              try {
+                connect(sourceInfo, connectCompileOptions, left_sub, right, context_mod)
+              } catch {
+                case BiConnectException(message) => throw BiConnectException(s".$field$message")
+              }
+            }
+        }
+      case (DontCare, right_r: Record) =>
```
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/BiConnect.scala]: There seems to be a lot of code duplication here, which might be better off in a separate function that gets called, between this and the Record-Record connection? Similarly with Vec, though to a lesser degree.

This isn't a huge issue, though.

Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/BiConnect.scala]: Why can DontCare be connected with a Record on either side, but the same isn't true with Vec?

File: chiselFrontend/src/main/scala/chisel3/internal/Builder.scala [hunk 28]
```diff
@@ -71,7 +71,11 @@ trait InstanceId {
 private[chisel3] trait HasId extends InstanceId {
   private[chisel3] def _onModuleClose: Unit = {} // scalastyle:ignore method.name
   private[chisel3] val _parent: Option[BaseModule] = Builder.currentModule
-  _parent.foreach(_.addId(this))
+  // Ignore null parents.
```
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/internal/Builder.scala]: What are the implications here of this potentially hiding bugs in the None case, since it will fail silently? Is there a way to make it fail noisily?

Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/internal/Builder.scala]: Actually, the new behavior isn't a "failure". It's required so the initial `DontCare` element (created in the package object) isn't added to the parent module's list of elements. Should we check `this` and fail noisely if `this` isn't a `DontCare`? Or provide a `noParent` method which is called in cases where we expect a parent but don't find one. The default method would complain, but `DontCare` (or any other implementation object) would override it to return without complaint.

Comment (ducky64) [reply chiselFrontend/src/main/scala/chisel3/internal/Builder.scala]: Sure, but only for the case of those singleton objects, right? If we encountered an issue with a non-singleton object where the parent was corrupted / misinitialized / not initialized, this seems like it would produce a failure that would silent discard useful data, or one that produces an exception elsewhere.

File: chiselFrontend/src/main/scala/chisel3/core/BiConnect.scala (hunk 27 above)
Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/BiConnect.scala]: I tried, but couldn't come up with anything that didn't duplicate the existing logic without adding additional code.

Comment (ucbjrl) [reply chiselFrontend/src/main/scala/chisel3/core/BiConnect.scala]: Short-sightedness. Will fix.

Comment (ducky64) [reply chiselFrontend/src/main/scala/chisel3/core/BiConnect.scala]: Yeah, that's fine, I couldn't come with anything obvious in the 2 minutes or so I thought about it.
